### PR TITLE
Optimize the max score tracking in the Query Phase of Hybrid Search

### PR DESCRIPTION
**Pull request by vibrantvarun**
### Description
Optimize the max score tracking in the QueryPhase of Hybrid Search. 
**Latency reduced**
```
(Number of Shards) x (Number of topDocs found on each shard)
```

### Issues Resolved
[764](https://github.com/opensearch-project/neural-search/issues/764)

### Check List
- [ ] New functionality includes testing.
    - [ ] All tests pass
- [ ] New functionality has been documented.
    - [ ] New functionality has javadoc added
- [ ] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java
Comment (shatejas) [inline src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java:120]: nit: ` maxScore = Math.max(score, maxScore)` for readability

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (shatejas) [inline src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: nit: randomness can introduce flakiness in the test. avoid if possible, seems simple enough to add two separate cases. Try and model your tests to be parametrized so you can easily add cases w/o copy pasting

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (shatejas) [inline src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:569]: any possibility we can introduce multiple leaves in the test? The nature of the collector is to collect top results across leaves. And the tests should ideally reflect that behavior for it to be tight. See if you can have a case where you can have at least one more leaf

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:569]: The [collect](https://github.com/opensearch-project/neural-search/pull/765/files#diff-efec3f3ef0d15cf4737a062d0f2b229eb984463e066013395a9a5b7e1867a40eL94) method runs for every doc Id found as a result of the query on a particular shard. What you are expecting is a step which is way before this method to execute. 

Let me explain how this thing works

1) Initialize leaf collectors from all the collectors at a shard level. A leaf collector is unique entity.

For example
Let say there are 2 collectors TopDocCollector and HybridTopScoreDocCollector. These 2 collectors will have a unique leaf collectors of there own. Those leaf collectors will run once per shard. 

2) Then execute those leaf collectors (In this case https://github.com/opensearch-project/neural-search/pull/765/files#diff-efec3f3ef0d15cf4737a062d0f2b229eb984463e066013395a9a5b7e1867a40eR51)
This will like
leafCollector.collect (docId1) ---> this call will be made from OpenSearch
leafCollector.collect (docId2)
leafCollector.collect (docId3)

docId1, docId2, docId3 are the results of HybridQuery
3) return the result to coordinator node.

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: I am following the pattern above which we used earlier to add tests in this Class. Btw, it does not matter if we use random in this case or not. Because the core logic we are testing here is the max score calculated by previous logic and the max score calculated by new logic. If at any point the two scores doesn't match then that will be considered as failure not flakiness.

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: If we have to change it to modularized or maybe optimize the tests then that has to be done in the entire project, Probably we cater it in one complete issue. The score of this PR is to just validate the max score logic.

File: src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java:120]: Sure I will add the change.

**Comment by VijayanB** [comment]
@vibrantvarun Do you have comparison in latency before and after this change?

**Comment by vibrantvarun** [comment]
> @vibrantvarun Do you have comparison in latency before and after this change?

Sure, I will add the latency graph. I am having some issues in the setup. Probably, once I resolve the issues I can add that flame graph in the issue of this PR. But, to unblock me on to raise my Sorting PR this PR needs to be merged.

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (shatejas) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: >  am following the pattern above 
I would highly encourage to break anti-patterns when you see one.

> Btw, it does not matter if we use random in this case or not.
Lets remove it then? pick a value

> If we have to change the tests to be parameterized then that has to be done in the entire project
You don't have to revamp all tests. its hard to justify the return on it. You can always start with new tests though.

File: src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java
Comment (VijayanB) [inline src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java:120]: nit: May be move this close to score variable

Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java:120]: The motive here is to compare the score which is added in the scoredoc with the maxScore. That is the  reason I have added it after the currentDoc. I can change this condition to currentDoc.score()

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: The thing is the place where random() is used it is lucene method. And lucene encourages to use randomized tests https://lucene.apache.org/core/7_3_1/test-framework/org/apache/lucene/util/LuceneTestCase.html . The reason is that it is not possible to cover all the test cases by using parameterized approach. Therefore, lucene recommends Randomized tests to increase more coverage.

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: At any other place which is not a lucene method we have used the paramertized approach for example DOCID1, DOCID2 etc.

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: https://github.com/randomizedtesting/randomizedtesting/wiki/Core-Concepts

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: Therefore, using random() is not a anti pattern.

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (shatejas) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:569]: While there is a new leafCollector, [you still are updating the priority queue and max score](https://github.com/opensearch-project/neural-search/pull/765/files#diff-efec3f3ef0d15cf4737a062d0f2b229eb984463e066013395a9a5b7e1867a40eL41) across collectors because its a member variable. And thats where the nature of this collector comes that is to collect the top docs across leaves and its important to test it out.

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (VijayanB) [inline src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:622]: seems you are depending on objects that are not part of testing. If those dependencies introduce any bug, this test will fail not because of what you want to test. Hence, either mock those external dependencies or convert to integration test if you feel you need to cover the complete flow.

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:622]: I am removing this unit test and will add a integ test instead. It will also address @shatejas and @VijayanB concerns.

File: src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java
Comment (shatejas) [reply src/test/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollectorTests.java:573]: Thanks for the article! 

Randomized testing seems to be having a use case when the complexity of the code is high and its not possible to cover cases everytime. Justifying a random input when it doesn't matter doesn't seems to be the right approach. I will leave it up to you how you want to keep this as it wasn't a blocker in the first place

**Review by VijayanB — APPROVED** [review]

**Review by junqiu-lei — APPROVED** [review]

**Review by jmazanec15 — COMMENTED** [review]
> (Number of Shards) x (Number of topDocs found on each shard)

Id be interested to see end user impact on this. topDocs will be a max of 10k right?

**Comment by martin-gaievski** [comment]
Final max score is calculated at the coordinator level as part of [normalization processor workflow](https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java#L113-L118). I think we can skip it calculation at shard level, maybe put 0.0. 
@vibrantvarun can you check this theory, my concern is that 0.0 score doesn't trigger any unwanted code path in core logic, plus check edge cases?

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryIT.java
Comment (shatejas) [inline src/test/java/org/opensearch/neuralsearch/query/HybridQueryIT.java:211]: The primary concern in the unit test was the behavior of collector during multi segment scenario. Does this initialization make sure of 2 segments at least

Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryIT.java:211]: We mostly write multiple segments scenarios when a results has been impacted by concurrent search is enabled and disabled. In this case the results does not get impacted by that.

**Pull request merged**
